# LOGISTIC REGRESSION MODEL WITH SKLEARN

Before I can use sklearn to build a model, I need to prep the data with Pandas. 

In [1]:
import pandas as pd

df = pd.read_csv('https://sololearn.com/uploads/files/titanic.csv')
#show 5 first data
df.head()

Survived  Pclass     Sex   Age  Siblings/Spouses  Parents/Children     Fare
0         0       3    male  22.0                 1                 0   7.2500
1         1       1  female  38.0                 1                 0  71.2833
2         1       3  female  26.0                 0                 0   7.9250
3         1       1  female  35.0                 1                 0  53.1000
4         0       3    male  35.0                 0                 0   8.0500

Before I can use sklearn to build a model, I need to prep the data with Pandas. 

In [3]:
df['Male'] = df['Sex'] == 'male'

Survived  Pclass     Sex   Age  Siblings/Spouses  Parents/Children  \
0           0       3    male  22.0                 1                 0   
1           1       1  female  38.0                 1                 0   
2           1       3  female  26.0                 0                 0   
3           1       1  female  35.0                 1                 0   
4           0       3    male  35.0                 0                 0   
..        ...     ...     ...   ...               ...               ...   
882         0       2    male  27.0                 0                 0   
883         1       1  female  19.0                 0                 0   
884         0       3  female   7.0                 1                 2   
885         1       1    male  26.0                 0                 0   
886         0       3    male  32.0                 0                 0   

        Fare   Male  
0     7.2500   True  
1    71.2833  False  
2     7.9250  False  
3    53.1000  False  
4     8.0500   True  
..       ...    ...  
882  13.0000   True  
883  30.0000  False  
884  23.4500  False  
885  30.0000   True  
886   7.7500   True  

[887 rows x 8 columns]

After that, I'll take all the features and create a numpy array called X. I first select all the columns I am interested in and then use the values method to convert it to a numpy array. Then I'll take the target (the Survived column) and store it in a variable y.

In [16]:
X = df[['Pclass', 'Male', 'Age', 'Siblings/Spouses', 'Parents/Children', 'Fare']].values
y = df['Survived'].values

## <span style="color:lightgreen">BUILD A LOGISTIC REGRESSION MODEL WITH SKLEARN</span>

In [8]:
from sklearn.linear_model import LogisticRegression

All sklearn models are built as Python classes. I first instantiate the class.

In [10]:
model = LogisticRegression()

Now I use the fit method to build the model

In [17]:
model.fit(X,y)

LogisticRegression()

Fitting the model means using the data to choose a line of best fit. The coefficients can be seen with the coef_ and intercept_ attributes.

In [18]:
print(model.coef_, model.intercept_)

[[-1.13645577 -2.64409939 -0.04237384 -0.38746619 -0.09619821  0.00297188]] [5.08856844]


Now I can use the predict method to make predictions.

In [19]:
model.predict(X)

array([0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0,

The first passenger in the dataset is:<br>
[3, True, 22.0, 1, 0, 7.25] <br>
This means the passenger is in Pclass 3, are male, are 22 years old, have 1 sibling/spouse aboard, 0 parents/child aboard, and paid $7.25. Let’s see what the model predicts for this passenger. Note that even with one datapoint, the predict method takes a 2-dimensional numpy array and returns a 1-dimensional numpy array.

In [22]:
print(model.predict([[3, True, 22.0, 1, 0, 7.25]]))

[0]


> The result is 0, it means the model predicts that this passenger did not survive.

Let’s see what the model predicts for the first 5 rows of data and compare it to the target array.

In [23]:
print(model.predict(X[:5])) 
# [0 1 1 1 0]
print(y[:5])

[0 1 1 1 0]
[0 1 1 1 0]


I can get a sense of how good the model is by counting the number of datapoints it predicts correctly. This is called the accuracy score.

In [27]:
y_pred = model.predict(X)
y == y_pred
print((y == y_pred).sum())

714


> From the 887 datapoints, the model makes the correct prediction for 714 of them.

To get the percent correct I'll use the score method. The score method uses the model to make a prediction for X and counts what percent of them match y.

In [30]:
model.score(X,y)

0.8049605411499436

> The accuracy of the model is 80%